guest-js/index.ts

```diff
@@ -1,19 +1,117 @@
-import { invoke } from '@tauri-apps/api/core';
+import { invoke, addPluginListener } from '@tauri-apps/api/core';
 import { listen, UnlistenFn } from '@tauri-apps/api/event';
 
-export class Download implements DownloadRecord {
+/**
+ * Manages subscriptions to download events from Rust and
+ * mobile plugins (iOS/Android), and dispatching these events
+ * to registered listeners.
+ */
+export class DownloadEventManager {
+   public static shared: DownloadEventManager = new DownloadEventManager();
+   private _listeners: Map<string, Set<(download: Download) => void>> = new Map();
+   private _eventUnlistenFn: UnlistenFn | null = null;
+   private _pluginListener: { unregister: () => void } | null = null;
+
+   private constructor() { }
+
+   /**
+    * Adds a listener for download events.
+    * @param key - The key of the download item to listen for.
+    * @param listener - The callback function to invoke when the download changes.
+    * @returns A promise with a function to remove this specific listener.
+    */
+   public async addListener(key: string, listener: (download: Download) => void): Promise<() => void> {
+      await this.ensureGlobalListeners();
+
+      if (!this._listeners.has(key)) {
+         this._listeners.set(key, new Set());
+      }
+
+      const listenersForKey = this._listeners.get(key);
+
+      if (listenersForKey) {
+         listenersForKey.add(listener);
+      }
+
+      // Return a function to remove this specific listener
+      return () => {
+         const listeners = this._listeners.get(key);
+
+         if (listeners) {
+            listeners.delete(listener);
+
+            // If no more listeners for this key, remove the key from the map.
+            if (listeners.size === 0) {
+               this._listeners.delete(key);
+            }
+         }
+
+         this.cleanupGlobalListeners();
+      };
+   }
+
+   private async ensureGlobalListeners(): Promise<void> {
```

Comment:
Could you make sure any `private`/`protected` methods in this MR start with an `_` / match [our naming standards](https://github.com/silvermine/silvermine-info/blob/master/coding-standards.md#naming-1)? I'm not sure why that's not enforced via ESLint, but I alerted the maintainers of our config and they're working on it.

Reply:
Good catch. Fixed. Thanks for following this up to with ESLint.